### PR TITLE
Implement HTTP Watch handler

### LINKED ISSUE #3305: Feature request: HTTP Endpoint for WatchHandler
Currently when a new watch is triggered, you can specify a handler that [execute an executable](https://github.com/hashicorp/consul/blob/master/agent/watch_handler.go)
 when changes are made. The executable get the trigger message through a JSON object on stdin. It would be nice to have an additional handler that specify an HTTP endpoint, similar to health checks.

As for an example configuration like this:
```
{
  "type": "key",
  "key": "foo/bar/baz",
  "http": "https://localhost:5000/watch",
}
```
the WatchHandler would send this when triggered:
```
POST /watch HTTP/1.1
Host: localhost:5000
Content-Type: application/json
Content-Length: 140
X-Consul-Index: 123

{
  "Key": "foo/bar/baz",
  "CreateIndex": 1793,
  "ModifyIndex": 1793,
  "LockIndex": 0,
  "Flags": 0,
  "Value": "aGV5",
  "Session": ""
}
```

The connection could `keep-alive` for a long time, if the server is agreeing on it.

Inspiration: https://developers.google.com/drive/v3/web/push
I would be willing to implement this.

### PR DESCRIPTION
**Pull request by HadarGreinsmark**
This makes it possible to be notified for updates using HTTP requests to a chosen endpoint. When a watch is triggered, a HTTP request is done with JSON payload. Before, it could only notify changes by executing an executable.

Resolves #3305

Todo before done:
- Make request configurable
- Fix tests
- HTTPS support

### CODE REVIEW

**Comment by magiconair** [comment]
This doesn't look right. Can you pls rebase your branch to master and force push so that we only see the changes?

**Comment by HadarGreinsmark** [comment]
@preetapan @magiconair Now I have implemented and tested it, so it would be great to get some feedback on it! Some tests are failing, but they don't seem to be related to this extension if I haven't missed anything. 

The watch request is configured directly in the watch definition as I think this is more logical and similar to check definitions. Currently, it doesn't reuse any connection with keep-alives as I don't think it's that important.

A lot of code is dedicated to parsing, as the parameters are given in the untyped Exempt variable. It would also be possible to parse all parameters using the mapstructure library and put that in a new Exempt parameter, but then a new struct has to be created and some manual checks still has to be done.

**Comment by HadarGreinsmark** [comment]
Do you have any thoughts regarding this @preetapan @magiconair ? If you agree on the configuration structure and Interface, I could start adding documentation in the pull request.

**Comment by preetapan** [comment]
@HadarGreinsmark sorry for the late response. 

Did a first pass through - Exempt has become a catch all for everything, and PR-3509 also adds an additional arg for subprocess shell handling.  

To make configuring this cleaner, I suggest adding a new handler type string to the [Plan struct here](https://github.com/hashicorp/consul/blob/d25be8671ad74a9af39a996c36d92eab6d0b1647/watch/watch.go#L13), and parsing all the additional http handler config options in another http_handler_config substructure. You would parse the additional args only when handler_type="http". Then that should be enough to create a http handler func instead of a shell handler func in agent.go, by using the handler type in the Plan struct.  

Basically, to configure a http handler we would do something like:
```
{
  "type": "key",
  "key": "testkey",
  "handler_type": "http",
   "http_handler_config": {
                                        "path":"http://myendpoint.com/foo"
                                         "method":"post"
                                         "timeout":"5s"
                                       ...
                                        }
}
```

**Comment by HadarGreinsmark** [comment]
@preetapan I agree that the Exempt variable is not the best solution. I can try to implement your config proposal. Though it may not be consistent with how http health checks are configured. Instead of Exempt, one solution would be to have a large struct for parsing all possible parameters and two specific structs; one for executables and another for http that both are a subset of the large struct. Similar to [CheckType](https://github.com/hashicorp/consul/blob/master/agent/structs/check_type.go) and [CheckHTTP](https://github.com/hashicorp/consul/blob/master/agent/check.go#L282). What do you think? I'm fine with any of the solutions.

**Comment by preetapan** [comment]
Configs for executables are changing a bit in https://github.com/hashicorp/consul/pull/3509 so would have to reconcile that with your changes. That's still in progress. 

Would prefer not to make more breaking config changes after 1.0 so adding another way to configure executables after that is not super ideal. Since HTTP watch handlers are a new concept we can have them configured the nicer newer way with a new config struct and leave the other one as is. That seems like the quickest path to getting this merged.

**Comment by HadarGreinsmark** [comment]
@preetapan Ok. I'll start with implementing this now.

**Comment by HadarGreinsmark** [comment]
@preetapan Now I have  re-implemented the config parsing according to the new format. So it is ready for a code review.

**Comment by HadarGreinsmark** [comment]
@preetapan have you some time to look into it? Added some additional fixes also.

File: watch/watch.go [hunk 1]
```diff
@@ -62,12 +75,33 @@ func ParseExempt(params map[string]interface{}, exempt []string) (*Plan, error)
...
+			return nil, fmt.Errorf("Handler type 'http' requires 'http_handler_config' to be set")
+		}
+		config, err := parseHttpHandlerConfig(params["http_handler_config"])
+		if err != nil {
+			return nil, fmt.Errorf(fmt.Sprintf("Failed to parse 'http_handler_config': %v", err))
+		}
+		plan.Exempt["http_handler_config"] = config
+		delete(params, "http_handler_config")
+
+	case "":
```
Comment (preetapan) [inline watch/watch.go]: Once you convert empty to "script" before this is ever called, you can turn this into "script"

File: agent/agent.go [hunk 2]
```diff
@@ -563,11 +563,11 @@ func (a *Agent) reloadWatches(cfg *config.RuntimeConfig) error {
 		} else if hasArgs && !ok {
 			return fmt.Errorf("Watch args must be a list of strings")
 		}
-		if hasHandler && hasArgs {
```
[the remark this answers is not in the document]
Comment (preetapan) [reply agent/agent.go]: Also, if the handler type is empty this can set it to "script" before ParseExempt is called. That way the parser layer only knows about two valid types "http" and "script" and doesn't support blank or a default case.

File: watch/watch.go [hunk 3]
```diff
@@ -129,3 +162,29 @@ func assignValueBool(params map[string]interface{}, name string, out *bool) erro
...
+	}
+
+	if config.Path == "" {
+		return nil, fmt.Errorf("Requires 'path' to be set")
+	}
+	if config.Method == "" {
+		config.Method = "POST"
+	}
+	if config.TimeoutRaw == "" {
+		config.Timeout = 10 * time.Second
```
Comment (preetapan) [inline watch/watch.go]: make the timeout a const defined in this file at the top.

**Comment by preetapan** [comment]
@HadarGreinsmark Sorry for the late response, busy with the 1.0 release and some fallout from that. I gave it an initial read and added some review comments now.

File: agent/watch_handler.go [hunk 4]
```diff
@@ -87,3 +90,72 @@ func makeWatchHandler(logOutput io.Writer, handler interface{}) watch.HandlerFun
...
+			}
+		} else {
+			trans.TLSClientConfig.InsecureSkipVerify = config.TLSSkipVerify
+		}
+
+		// Create the HTTP client.
+		httpClient := &http.Client{
+			Timeout:   config.Timeout,
+			Transport: trans,
+		}
```
Comment (preetapan) [inline agent/watch_handler.go]: Use context for proper cancellation of requests that have timed out.  Create a context [WithTimeout](https://golang.org/pkg/context/#WithTimeout), and wire that into the http request below via https://golang.org/pkg/net/http/#Request.WithContext

File: agent/watch_handler.go [hunk 5]
```diff
@@ -87,3 +90,72 @@ func makeWatchHandler(logOutput io.Writer, handler interface{}) watch.HandlerFun
...
+		// Get the output, add a message about truncation
+		outputStr := string(output.Bytes())
+		if output.TotalWritten() > output.Size() {
+			outputStr = fmt.Sprintf("Captured %d of %d bytes\n...\n%s",
+				output.Size(), output.TotalWritten(), outputStr)
+		}
+
+		if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
+			// Log the output
+			logger.Printf("[DEBUG] agent: http watch handler '%s' output: %s", config.Path, outputStr)
```
Comment (preetapan) [inline agent/watch_handler.go]: Make this trace rather than debug.

**Comment by HadarGreinsmark** [comment]
@preetapan now your comments are fixed.

File: agent/agent.go [hunk 6]
```diff
@@ -534,7 +534,13 @@ func (a *Agent) reloadWatches(cfg *config.RuntimeConfig) error {
 	// Compile the watches
 	var watchPlans []*watch.Plan
 	for _, params := range cfg.Watches {
-		// Parse the watches, excluding the handler
+		if handlerType, ok := params["handler_type"]; !ok {
+			params["handler_type"] = "script"
+		} else if handlerType != "http" && handlerType != "script" {
+			return fmt.Errorf(fmt.Sprintf("Handler type '%s' not recognized", params["handler_type"]))
```
Comment (preetapan) [inline agent/agent.go]: This shouldn't need the unnecessary inner `fmt.Sprintf(..)

Comment (HadarGreinsmark) [reply agent/agent.go]: Fixed

File: agent/agent.go [hunk 7]
```diff
@@ -534,7 +534,13 @@ func (a *Agent) reloadWatches(cfg *config.RuntimeConfig) error {
 	// Compile the watches
```
Comment (preetapan) [inline agent/agent.go]: @HadarGreinsmark This is ready to merge, can you also add a section describing the new http handler to https://github.com/hashicorp/consul/blob/master/website/source/docs/agent/watches.html.md and I can take it from there after that. 
Thanks for working on this.

**Comment by HadarGreinsmark** [comment]
@preetapan now some documentation is added. Also added small fix: to send JSON Content-Type to requests.
